pkg/controller/scansettingbinding/scansettingbinding_controller.go

```diff
@@ -0,0 +1,515 @@
+package scansettingbinding
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	compliancev1alpha1 "github.com/openshift/compliance-operator/pkg/apis/compliance/v1alpha1"
+	"github.com/openshift/compliance-operator/pkg/controller/common"
+	"github.com/openshift/compliance-operator/pkg/utils"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"strings"
+	"time"
+)
+
+const (
+	// The default time we should wait before requeuing
+	requeueAfterDefault = 10 * time.Second
+)
+
+var log = logf.Log.WithName("controller_scansettingbinding")
+
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileScanSettingBinding{client: mgr.GetClient(), scheme: mgr.GetScheme(), recorder: mgr.GetEventRecorderFor("scansettingbindingctrl")}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("scansettingbinding-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource ScanSettingBinding
+	err = c.Watch(&source.Kind{Type: &compliancev1alpha1.ScanSettingBinding{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to secordary resource ScanSetting. Since Setting does not link directly to a Binding,
+	// but the other way around, we use a mapper to enqueue requests for Binding(s) used by a Setting
+	err = c.Watch(&source.Kind{Type: &compliancev1alpha1.ScanSetting{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: &scanSettingMapper{mgr.GetClient()},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileScanSettingBinding implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileScanSettingBinding{}
+
+// ReconcileScanSettingBinding reconciles a ScanSettingBinding object
+type ReconcileScanSettingBinding struct {
+	client   client.Client
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+// FIXME: generalize for other controllers?
```

Comment:
That would be a nice refactoring for another PR.

At this point, it would be nice to have a common Reconcile object that allows us to set and use common things such as the event recorder and logging with a simple function.

Reply:
So far I added a new events.go module with a safer even that can or doesn't have to be initialized (unit tests can't initialize events it seems). If you prefer to also have a common reconcile object, I think it is out of scope of this PR.

Reply:
yep, that's why I said that it would be for another PR.